libp2p/protocols/discovery/rendezvous.nim

```diff
@@ -0,0 +1,348 @@
+  let
+    n = now()
+    nsSalted = ns & rdv.salt
+  discard rdv.indexes.hasKeyOrPut(nsSalted, initOffsettedSeq[int]())
+  try:
+    for index in rdv.indexes[nsSalted]:
+      if rdv.registered[index].peerId == peerId:
+        rdv.registered[index].expireDate = rdv.defaultDT
+    rdv.registered.add(
+      RegisteredData(
+        peerId: peerId,
+        expireDate: now() + initDuration(ttl.int64),
+        data: r
+      )
+    )
+    rdv.indexes[nsSalted].add(rdv.registered.high)
+#    rdv.registerEvent.fire()
+  except KeyError:
+    doAssert false, "Should have key"
+
+proc register(rdv: RendezVous, conn: Connection, r: Register): Future[void] =
+  let ns = r.ns.get("")
+  if ns.len notin 1..255:
+    return conn.sendRegisterResponseError(InvalidNamespace)
+  let ttl = r.ttl.get(MinimumTTL)
+  if ttl notin MinimumTTL..MaximumTTL:
+    return conn.sendRegisterResponseError(InvalidTTL)
+  let pr = checkPeerRecord(r.signedPeerRecord, conn.peerId)
+  if pr.isErr():
+    return conn.sendRegisterResponseError(InvalidSignedPeerRecord, pr.error())
+  if rdv.countRegister(conn.peerId) >= RegistrationLimitPerPeer:
+    return conn.sendRegisterResponseError(NotAuthorized, "Registration limit reached")
+  rdv.save(ns, conn.peerId, r, ttl)
+  conn.sendRegisterResponse(ttl)
+
+proc unregister(rdv: RendezVous, conn: Connection, u: Unregister) =
+  let nsSalted = u.ns.get("") & rdv.salt
+  try:
+    for index in rdv.indexes[nsSalted]:
+      if rdv.registered[index].peerId == conn.peerId:
+        rdv.registered[index].expireDate = rdv.defaultDT
+  except KeyError:
+    return
+
+proc discover(rdv: RendezVous, conn: Connection, d: Discover): Future[void] =
+  let ns = d.ns.get("")
+  if ns.len notin 0..255:
+    return conn.sendDiscoverResponseError(InvalidNamespace)
+  var
+    limit = min(DiscoverLimit, d.limit.get(DiscoverLimit))
+  let
+    nsSalted = ns & rdv.salt
+    cookie = if d.cookie.isSome():
+      try: Protobuf.decode(d.cookie.get(), type(Cookie))
+      except ProtobufReadError: return conn.sendDiscoverResponseError(InvalidCookie)
+    else: Cookie(offset: rdv.registered.low().uint64, ns: ns)
+    indexes = if ns != "":
+        try: rdv.indexes[nsSalted].s
+        except KeyError: return conn.sendDiscoverResponseError(InvalidNamespace)
+      else: toSeq(rdv.registered.low()..rdv.registered.high())
+  if indexes.len() == 0:
+    return conn.sendDiscoverResponse(@[], Cookie())
+  var offset = indexes[^1]
+  let s = collect(newSeq()):
+      for index in indexes:
+        if index.uint64 < cookie.offset: continue
+        if limit == 0:
+          offset = index
+          break
+        limit.dec()
+        rdv.registered[index].data
+  conn.sendDiscoverResponse(s, Cookie(offset: offset.uint64, ns: ns))
+
+proc advertisePeer(rdv: RendezVous, peer: PeerId, msg: seq[byte]) {.async.} =
+  let conn = await rdv.switch.dial(peer, RendezVousCodec)
+  defer: await conn.close()
+  await conn.writeLp(msg).wait(chronos.milliseconds(2500))
```

Comment:
This will raise an exception on timeout, but we are in a asyncSpawned context, so any uncaught exception will be raise as a defect

Another option might be to start a `let timeout = sleepAsync(5.seconds)` at the beginning of the procedure, and do
```nim
let writing = conn.writeLp(msg)
await writing or timeout
if timeout.finished:
  writing.cancel()
  return
```